### ISSUE TITLE
Logger: Log Basic Information

### ISSUE THREAD

**Issue by bastianilso** [med-material/Whack_A_Mole_VR#2]
We need a logger which can log important events in the Whack-A-Mole game.
The logger should produce a CSV file, where each row is a game event, fx one of the following:
- [x] "Game Started"
- [x] "Target Hit"
- [x] "Target Missed"
- [x] "Wrong Target Hit"
- [x] "Game Over"

For each event, we should also log the following:
- [x] Log a Timestamp (System.DateTime.Now).
- [x] Log a Test Number ("TestNo") and Participant ID ("PID") so we can distinguish tests from each other, when several are performed for the same participants and in-between participants.
- [x] Log how much time passed since the last event ("TimeSinceLastEvent")
- [x] Log a comment which is a custom string (comment) set to a participant.
- [x] Log TargetHitByPlayer which logs what target ID has been hit by the player.
- [x] Log CurrentTargetToHit which log which target the player was supposed to hit.

- [x] Log TargetHitByPlayerPositionWorld as a world position of the target the player hit (transform.position).
- [x] Log TargetHitByPlayerPositionLocal as a local position of the target the player hit (transform.localPosition).
- [x] Log CurrentTargetToHitPosWorld which log which target the player was supposed to hit (transform.position).
- [x] Log CurrentTargetToHitPosLocal which log which target the player was supposed to hit (transform.localPosition).

**Comment by QuentinDaveau** [comment]
To do:

- [ ]  Add unique Id generation (participantID+testNumber+date?);

- [ ] Log mole whacking reaction time.

**Comment by QuentinDaveau** [comment]
To do:
- [ ] Add game duration

- [ ] Add game time spent

- [ ] Add game time left

**Comment by QuentinDaveau** [comment]
Related to issue #7.

**Comment by QuentinDaveau** [comment]
@bastianilso 
Currently the logger logs:

- A timestamp, the time since last event, a unique ID for each game, the event raised, the state of each modifier, the participant and test ID, the game state, the game duration, the game speed (difficulty), the game time spent, the game time left

- If the event concerns a Mole (mole pop, mole missed, mole spawned) are also logged the position (local and world) of the Mole, the lifetime of the Mole, the lifetime spent in case of Mole pop (reaction time), the ID of the Mole, the target mole local and world position, and the tracked objects (meaning the controllers, the laser, and the head, see #7) position, rotation, position travelled and rotation travelled.

The events triggering these logs are:

Game state change (start, stop, finished, pause, unpause), Mole interaction (spawned, fake spawned, missed, hit, wrong hit), any change to the game settings (difficulty, duration) and any change to the modifiers (dual task, prism effect...)

Here is a sample of the resulting logs:

[newLog2_29-10-2019_12-01-37.zip](https://github.com/med-material/Whack_A_Mole_VR/files/3783196/newLog2_29-10-2019_12-01-37.zip)

Any suggestion (more things to log, things to change...) before I create a pull request ?

**Comment by bastianilso** [comment]
@QuentinDaveau : from a quick look, here is a few things i noticed

- the csv file starts with `sep=;` in line 1.
- the time stamp contains an underscore. I suggest you use `System.DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.ffff");`
- PrismEffect logs with a comma `,` while the TimeSinceLastEvent logs with a period `.` mark. I would suggest we stick to using period marks `.` since thats the default in other projects.
- It is not clear to me what `TrackPosWorldX0`, `TrackObject0` etc. are supposed to be for. Could you explain?
- I dont see any columns called `TargetHitByPlayerPositionWorld`, ie. `TargetHitByPlayerPositionWorldX`, `TargetHitByPlayerPositionWorldY`, `TargetHitByPlayerPositionWorldZ` - same for `CurrentTargetToHitPosWorld`
- The column `CurrentTargetToHit` contains numbers that I am not sure how to understand. Maybe we can use an enum here whih sensibly names the moles in a logical way. Fx `Mole(1,1)` Could be the top left mole, `Mole(2,1)` could be the mole. We might want to have `MoleIndexX` and `MoleIndexY` to additionally keep the X-Y indexes easily available in data analysis.

It might be good if you could also update the readme file with explanations of the columns and what they are supposed to log. See for instance [the Readme file in VREyeTrackerAccuracy](https://github.com/med-material/VREyeTrackingAccuracyTest).

**Comment by QuentinDaveau** [comment]
@bastianilso thanks for the feedback. Concerning the remarks:

> the csv file starts with sep=; in line 1.

It is to specify to any software that would read the csv what is the separator. In this case, it is ";".

> the time stamp contains an underscore. I suggest you use System.DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.ffff");

That's true, it also fit with the datetime entry in SQL. I'll replace the underscore with a space.

> PrismEffect logs with a comma , while the TimeSinceLastEvent logs with a period . mark. I would suggest we stick to using period marks . since thats the default in other projects.

That's a mistake. I'll fix it.

> It is not clear to me what TrackPosWorldX0, TrackObject0 etc. are supposed to be for. Could you explain?

It corresponds to the tracked object's properties. A track object is an object which position and/or rotation, position travelled and/or rotation travelled are logged (for issue #7). The column "TrackObject" refers to which object it is question of. So if you have TrackObject0 = Left Controller, you know that TrackObjectPosWorldZ0 refers to the left controller. I did it this way to prevent having names like "LeftController" as column, since we may want to track other objects in the future/ stop tracking certain objects and that may be an issue with SQL. 
If you want I can set them to "LeftControllerPosWorldX...".

> I dont see any columns called TargetHitByPlayerPositionWorld, ie. TargetHitByPlayerPositionWorldX, TargetHitByPlayerPositionWorldY, TargetHitByPlayerPositionWorldZ - same for CurrentTargetToHitPosWorld

TargetHitByPlayerPositionWorld / TargetHitByPlayerPositionLocal are missing simply because I didn't pop any Mole while logging. If I had pop any mole, the column would have been created. This is due to the way the logger works, the columns aren't hardcoded but created each time an event has an uknown parameter (like TargetHitByPlayerPosWorld). Once again, if you want I can do so every column is initialized right at the beginning, so even an empty column would be logged right at the start of the game. CurrentTargetToHitPositionWorld/Local is present in the CSV file, you might have missed it.

> The column CurrentTargetToHit contains numbers that I am not sure how to understand. Maybe we can use an enum here whih sensibly names the moles in a logical way. Fx Mole(1,1) Could be the top left mole, Mole(2,1) could be the mole. We might want to have MoleIndexX and MoleIndexY to additionally keep the X-Y indexes easily available in data analysis.

The CurrentTargetToHit column is actually quite close to what you said. It is the ID of the Mole, which is composed of 4 decimals. The first two references the X position (starting from 00), and the next two references the Y position. So a mole with an ID = 0105 would be on the second column and the sixth row. The first 0 are automatically removed when saving the log, that's why we have some ID like "3", which actually means "0003" (first colums, fourth row). I agree that's not the best. What I'll do is start from 1 instead of 0 (clearer to people not used to code), and add a MoleIndexX and MoleIndexY. I'll also try to force four digits, so we have "0101" instead of "101".

> It might be good if you could also update the readme file with explanations of the columns and what they are supposed to log. See for instance the Readme file in VREyeTrackerAccuracy.

I'll update the readme once the log format (columns name...) is globally defined.

**Comment by bastianilso** [comment]
@QuentinDaveau thanks for the explanations and reply, just a few remarks:

> It is to specify to any software that would read the csv what is the separator. In this case, it is ";".

hmm, well neither [LibreOffice](https://user-images.githubusercontent.com/3967945/67774654-0dfc0200-fa5e-11e9-8584-29155a0f47a4.png) or [R Studio](https://user-images.githubusercontent.com/3967945/67774414-9ded7c00-fa5d-11e9-8165-25e6536706a0.png) seems to like it. :confused:
I suggest for now we stick to the [RFC 4180 specification](https://tools.ietf.org/html/rfc4180#section-2).

```
field_name,field_name,field_name CRLF
aaa,bbb,ccc CRLF
zzz,yyy,xxx CRLF

```

> "TrackObject" refers to which object it is question of. So if you have TrackObject0 = Left Controller, you know that TrackObjectPosWorldZ0 refers to the left controller. I did it this way to prevent having names like "LeftController" as column, since we may want to track other objects in the future/ stop tracking certain objects and that may be an issue with SQL.
If you want I can set them to "LeftControllerPosWorldX...".

I understand, that's clever! However, in the data analysis step we can get into trouble/complications if objects start changing columns. For the sake of readability (and sensible auto-completion in R Studio), I would prefer if we gave column headers human legible names. Then what we can do in SQL when we remove/add columns, is to provide every column with a Default Value. In cases where the data is missing, we can then default to NULL, which is easy to program around in R Shiny.


>  Once again, if you want I can do so every column is initialized right at the beginning, so even an empty column would be logged right at the start of the game.

I would prefer that all columns are available in the CSV file, and that human-readable fallback values are logged in those cases, fx logging `NULL`, for `TargetHitByPlayerPositionWorldX` when no target is being hit. I know that sounds like a lot of unnecessary columns, but it is to create a sense of security / and to make it possible to run automated scripts on the data, where we can expect all columns to be present.

**Comment by QuentinDaveau** [comment]
> I would prefer that all columns are available in the CSV file, and that human-readable fallback values are logged in those cases, fx logging NULL, for TargetHitByPlayerPositionWorldX when no target is being hit. I know that sounds like a lot of unnecessary columns, but it is to create a sense of security / and to make it possible to run automated scripts on the data, where we can expect all columns to be present.

Should I leave an empty field when a columns doesn't have a value (in the CSV this would give ",,", this is what is currently done) or should I explicitely set "NULL" (so it would read ",NULL," in the CSV) ?

**Comment by QuentinDaveau** [comment]
@bastianilso 
Updated the logger to fit the remarks from the previous comments.

The new logs looks like this:

[newLog2_30-10-2019 14-27-20.zip](https://github.com/med-material/Whack_A_Mole_VR/files/3789170/newLog2_30-10-2019.14-27-20.zip)

Any remark/suggestion ?

**Comment by bastianilso** [comment]
@QuentinDaveau looks good!

very minor remarks:
- timestamp does not capture milliseconds? (i only see fx 30/10/2019 14:27:20 its missing `.ffff`)
- how do you generate the gameID?
- PositionWorldX, Y, Z - does it refer to PlayerPosition? in that case, i think adding Player in front might make it more clear (PlayerPositionWorldX etc).
- LifeTime..is that MoleLifeTime? :-)
- ActivatedDuration..is that MoleActivatedDuration? :-)

**Comment by QuentinDaveau** [comment]
@bastianilso ok :-), I'll add the milliseconds then ^^

Just thought about something, every column concerning the mole is recognized by the "Mole" keyword in its name ("MolePositionWorldX", "MoleLifetime"...). Shouldn't we also include the "Mole" keyword in the curent target columns for the sake of consistency, since it also refers to a Mole (so we would have "CurrentMoleToHit" instead of "CurrentTargetToHit"). I think it would be a bit clearer.

**Comment by QuentinDaveau** [comment]
Issue fixed in merging of branch #23. Closing the issue.
